### PR TITLE
TCP new features

### PR DESCRIPTION
**Pull request by jparisu**
Features included:
- TLS peer verification
- SNI support
- External port for TCP

Includes 
- #299 
- #307

### CODE REVIEW

File: resources/configurations/examples/wan_server.yaml [hunk 1]
```diff
@@ -34,9 +34,11 @@ participants:
...
   - name: WanParticipant                                          # 6
     kind: wan                                                     # 7
-    listening-addresses:                                          # 8
-      - ip: 1.1.1.1
-        port: 12666
+    listening-addresses:
+      - ip: 1.1.1.1                                               # 8
+        port: 11666                                               # 9
+        external-port: 11777                                      # 10
+        transport: tcp                                            # 11
```
Comment (juanlofer-eprosima) [inline resources/configurations/examples/wan_server.yaml:41]: I would leave the WAN examples in resources to use UDP transport.

File: docs/rst/user_manual/wan_configuration.rst [hunk 2]
```diff
@@ -210,7 +207,8 @@ User *B* should set its *listening-addresses* and *connection-addresses* as foll
         id: 3                             # Must be different than A one
       listening-addresses:
         - ip: 2.2.2.2                     # Public IP of host Hb
-          port: 11777                     # Port forwarded router Rb
+          port: 11778                     # This device physical port forwarded internally
+          external-port: 11777            # Port forwarded router Rb
```
Comment (juanlofer-eprosima) [inline docs/rst/user_manual/wan_configuration.rst:211]: I believe external port only applies to TCP transport.

**Review by juanlofer-eprosima — CHANGES_REQUESTED** [review]

File: docs/rst/user_manual/configuration.rst [hunk 3]
```diff
@@ -334,6 +336,16 @@ An Address is defined by:
     domain: localhost             # DNS call return value = 127.0.0.1
     port: 33333
 
+
+External Port
+-------------
+
+External port is used to configure a Server (Discovery Server or Initial Peers) that runs under a NAT.
+In case this value is not set, the external port is supposed to be the same as the internal one.
+In this case the external and internal port in the network router port forwarding must coincide.
```
[the remark this answers is not in the document]
Comment (jparisu) [reply docs/rst/user_manual/configuration.rst:345]: I tried to fix it

File: resources/configurations/examples/wan_ds_client.yaml [hunk 4]
```diff
@@ -46,7 +46,7 @@ participants:
       - ip: 2.2.2.2
         port: 11670                                               # 13
         external-port: 11680                                      # 14
-        transport: udp                                            # 15
+        transport: tcp                                            # 15
```
Comment (juanlofer-eprosima) [inline resources/configurations/examples/wan_ds_client.yaml:49]: This needs to be changed I believe. In WAN, having a listening and connection addresses made sense for udp, not with tcp. Also, the connection address is not pointing at the right port (and transport is not udp).

**Review by juanlofer-eprosima — CHANGES_REQUESTED** [review]

**Review by juanlofer-eprosima — APPROVED** [review]
LGTM

**Pull request merged**